Snooz integration supports Pro and Breez models

AustinBrunkhorst opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
- Updates pysnooz to [0.9.2](https://github.com/AustinBrunkhorst/pysnooz/compare/v0.8.6...v0.9.2)
- Officially adds support for Pro and Breez models

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
pysnooz [0.9.0](https://github.com/AustinBrunkhorst/pysnooz/releases/tag/v0.9.0) introduces breaking changes:
- `SnoozAdvertisementData` is now a dataclass and returned from `parse_snooz_advertisement` during discovery
- The config flow stores `model` and `firmware_version` to construct a `SnoozAdvertisementData` during setup
- During setup, we automatically update entries that are missing properties by searching for a supported advertisement.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #99231
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28859

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

AustinBrunkhorst committed:
test progress

AustinBrunkhorst committed:
add tests for migration

AustinBrunkhorst committed:
linter fixes

AustinBrunkhorst committed:
update pysnooz to 0.9.0

joostlek reviewed (changes requested):
This PR is too big. I think the 4 points you mention are a good way to divide the PRs into smaller PRs.

AustinBrunkhorst committed:
revert non-essential changes

joostlek commented:
Right, can we split this up even more? First do the dependency bump. Does that bump require changes? I'm also seeing a config flow update, is that for every device? Can we split that off as well?

AustinBrunkhorst commented:
> Right, can we split this up even more? First do the dependency bump. Does that bump require changes? I'm also seeing a config flow update, is that for every device? Can we split that off as well?

Unfortunately, this is the smallest this update can be to pick up 0.9.0.

> SnoozAdvertisementData is now a dataclass and returned from parse_snooz_advertisement during discovery.
> The config flow stores model and firmware_version to construct a SnoozAdvertisementData during setup.

Model detection is dependent on advertisement data, and the changes in this PR are what's minimally required to support this.

joostlek commented on homeassistant/components/snooz/config_flow.py:
```diff
@@ -31,11 +35,16 @@ class DiscoveredSnooz:
     """Handle a config flow for Snooz."""
 
-    VERSION = 1
+    VERSION = 2
```
You updated the version. This means you need a migration. I only glanced at the code so far (am traveling so I'm on my phone so I can't get a full picture) and what I understood is that we now store more options. We can also do that without a migration by adding the required information in the init (although I am currently reworking it, checkout the current withings integration. In `__init__.py` at the `async_setup_entry` extra info is gathered to update the config entry. Would this work?)

The downside of doing a migration is that it's not reversable. So if this gets in 2023.10, they can't just install 2023.9 and expect it to work, since the config entry has a different version. So although we have the whole migration session, we try to avoid it as much as possible, if it's not needed.

joostlek replied:
Oh I now see you change from pairing token to password? So then a migration can become useful

AustinBrunkhorst replied:
pysnooz changed naming from pairing token to password to be consistent with Snooz's internal naming. In the snooz component, config data will remain `token` since it's just a semantic change.

Pre `0.9.0` only SNOOZ Original was supported. Model detection requires the advertised name with the specific format `Model-XXXX`. I see what you're saying about doing the migration logic. I will include it as part of the setup process instead of storing `model` and `password`.

Is it acceptable for `async_setup_entry` to modify data? One downside to moving the migration logic into `async_setup` is that it will increase setup time by default since often times cached advertisements use the device characteristic name instead of the name from an original advertisement. This is always `Snooz`, so it breaks model detection that requires `Mpdel-XXXX`. We can avoid the startup time by caching `model`.

AustinBrunkhorst committed:
fix test failures

AustinBrunkhorst committed:
advertisement information is loaded as nessecary during setup

AustinBrunkhorst committed:
Merge branch 'dev' into pysnooz_0_9_0

AustinBrunkhorst committed:
Merge branch 'dev' into pysnooz_0_9_0

AustinBrunkhorst committed:
update pysnooz to 0.9.2

AustinBrunkhorst committed:
update requirements

AustinBrunkhorst committed:
Merge branch 'dev' into pysnooz_0_9_0

AustinBrunkhorst committed:
Merge branch 'dev' into pysnooz_0_9_0

AustinBrunkhorst committed:
Merge branch 'dev' into pysnooz_0_9_0

AustinBrunkhorst committed:
Merge branch 'dev' into pysnooz_0_9_0

AustinBrunkhorst committed:
Merge branch 'dev' into pysnooz_0_9_0

AustinBrunkhorst committed:
Merge branch 'dev' into pysnooz_0_9_0

AustinBrunkhorst committed:
Merge branch 'dev' into pysnooz_0_9_0

AustinBrunkhorst committed:
Merge branch 'dev' into pysnooz_0_9_0

AustinBrunkhorst commented:
@joostlek Is the PR size reasonable now, given the required changes from `pysnooz`? I can do the new `async_setup` tests in a separate PR but it will only bring it down ~70 lines

bdraco commented on homeassistant/components/snooz/__init__.py:
```diff
@@ -60,3 +61,103 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+) -> SnoozAdvertisementData | None:
+    """Process advertisements for an address until a supported advertisement is found."""
+
+    def is_supported_advertisement(
+        service_info: BluetoothServiceInfo,
+    ) -> bool:
+        if parse_snooz_advertisement(service_info) is None:
+            LOGGER.warning(
+                f"Skipping unsupported advertisement with name: {service_info.name}"
+            )
+            return False
+
+        return True
```
It looks like this doesn't need to be an inner function and can be moved outside this function.

bdraco commented on homeassistant/components/snooz/config_flow.py:
```diff
@@ -173,24 +177,33 @@ async def async_step_pairing_timeout(
-        device = self._discovery.device
 
         def is_device_in_pairing_mode(
             service_info: BluetoothServiceInfo,
         ) -> bool:
-            return device.supported(service_info) and device.is_pairing
+            adv_data = parse_snooz_advertisement(service_info)
+
+            if adv_data is None or not adv_data.is_pairing:
+                return False
+
+            assert self._discovery
+            # copy the password from this advertisement to the discovery object
+            self._discovery.device.password = adv_data.password
+
+            return True
```
It looks like this doesn't need to be an inner function either

bdraco commented on homeassistant/components/snooz/__init__.py:
```diff
@@ -60,3 +61,103 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    )
+
+
+async def _async_update_entry_config(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> SnoozAdvertisementData:
+    """Find a supported advertisement and store data in config entry."""
+
+    address = entry.data[CONF_ADDRESS]
+    LOGGER.debug(f"Loading device information from advertisement for address {address}")
+
+    adv_data = await hass.async_create_task(
+        _async_wait_for_advertisement(hass, address)
+    )
```
Can you split this function up and or rename it to better show intent. Its doing three things right now.

1. Updating the config entry
2. Waiting for an advertisement
3. Optionally extending the adv data from the config entry data

frenck commented:
I've marked this PR, as changes are requested that need to be processed.
Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)
